lib/moduleEncoding.js

```diff
@@ -0,0 +1,558 @@
+		}
+		if (hash.length != 32) {
+			throw new RangeError('Invalid size of the SHA-256 checksum');
+		}
+		hash.copy(data, offs);
+	}
+	offs += 32;
+	// Suffix size
+	offs = data.writeUInt16LE(size, offs);
+	// CRC-32 checksum
+	if (suffix.crcBlock !== undefined) {
+		let crc = suffix.crcBlock;
+		if (typeof crc === 'string') {
+			crc = Buffer.from(crc, 'hex');
+		}
+		if (crc.length != 4) {
+			throw new RangeError('Invalid size of the CRC-32 checksum');
+		}
+		crc.copy(data, offs);
+	}
+	offs += 4;
+	return data;
+}
+
+/**
+ * Compress the firmware module.
+ *
+ * @param {Buffer} data Module data.
+ * @param {Object} [prefix] Prefix info (see `HalModuleParser`).
+ * @param {Object} [suffix] Suffix info (see `HalModuleParser`).
+ * @param {Object} [options] Compression options (https://nodejs.org/api/zlib.html#zlib_class_options).
+ * @returns {Buffer}
+ */
+async function compressModule(data, prefix, suffix, options) {
+	if (!prefix || !suffix) {
+		const parser = new HalModuleParser();
+		const { prefixInfo, suffixInfo } = await parser.parseBuffer({ fileBuffer: data });
+		if (!prefix) {
+			prefix = prefixInfo;
+		}
+		if (!suffix) {
+			suffix = suffixInfo;
+		}
+	}
+	let startAddr = prefix.moduleStartAddy;
+	if (startAddr === undefined) {
+		throw new RangeError('moduleStartAddy is missing');
+	}
+	if (typeof startAddr === 'string') {
+		startAddr = Number.parseInt(startAddr, 16);
+	}
+	let endAddr = prefix.moduleEndAddy;
+	if (endAddr === undefined) {
+		throw new RangeError('moduleEndAddy is missing');
+	}
+	if (typeof endAddr === 'string') {
+		endAddr = Number.parseInt(endAddr, 16);
+	}
+	const flags = prefix.moduleFlags;
+	if (flags === undefined) {
+		throw new RangeError('moduleFlags is missing');
+	}
+	if (flags & ModuleFlags.COMPRESSED) {
+		throw new RangeError('Module is already compressed');
+	}
+	if (flags & ModuleFlags.COMBINED) {
+		// Combined modules need to be compressed individually
+		throw new RangeError('Can\'t compress a combined module');
+	}
+	const suffixSize = suffix.suffixSize;
+	if (suffixSize === undefined) {
+		throw new RangeError('suffixSize is missing');
+	}
+	// Some very old firmware binaries have a malformed suffix
+	if (suffixSize === 0xffff) {
+		throw new RangeError('Invalid suffix size');
+	}
+	let dataSize = endAddr - startAddr + 4 /* CRC-32 */;
+	const prefixOffs = prefix.prefixOffset || 0;
+	if (data.length < prefixOffs + MODULE_PREFIX_SIZE + suffixSize + 4 || data.length !== dataSize) {
+		throw new RangeError('Invalid size of the module data');
+	}
+	let dataOffs = 0;
+	if (flags & ModuleFlags.DROP_MODULE_INFO) {
+		if (prefixOffs !== 0) {
+			throw new RangeError('DROP_MODULE_INFO can\'t be set for a module with a vector table');
+		}
+		// Skip the prefix data
+		dataOffs = MODULE_PREFIX_SIZE;
+		dataSize -= MODULE_PREFIX_SIZE;
+	}
+	// Compress the module data
+	// TODO: Investigate compressing the module data in place to optimize memory usage, or use some
```

Comment:
Remove this comment. If you're not doing it now, I doubt anybody is going to come back and do this. It also feel like premature optimization compared to all the other memory allocations that are happening in device service. A more meaningful optimization would be to store the compressed binary back in the database for system firmware and product firmware such that the next device requesting a compressed binary will not trigger the compression operation in device service.

Reply:
Why? It's just a comment pointing out something that can be improved, not a request for action :slightly_smiling_face: While in-place compression is indeed a convoluted approach, using liked buffers makes sense in both scenarios, especially for combined modules. After all, we just need to send this data to the client, it doesn't matter whether the buffer is contiguous.